packages/ember-data/lib/serializers/json-serializer.js

```diff
@@ -274,11 +276,42 @@ export default Serializer.extend({
   normalizeErrors: function(typeClass, hash) {
     this.normalizeId(hash);
     this.normalizeAttributes(typeClass, hash);
-    this.normalizeRelationships(typeClass, hash);
+    this.normalizeRelationshipErrors(typeClass, hash);
     this.normalizeUsingDeclaredMapping(typeClass, hash);
   },
 
   /**
+    @method normalizeRelationshipErrors
+    @private
+  */
+  normalizeRelationshipErrors: function(typeClass, hash) {
+    if (!this.keyForRelationship) {
+      return;
+    }
+
+    typeClass.eachRelationship(function(key, relationship) {
+      var payloadKey = this.keyForRelationship(key, relationship.kind);
+      if (key !== payloadKey && hash.hasOwnProperty(payloadKey)) {
+        hash[key] = hash[payloadKey];
+        delete hash[payloadKey];
+      }
+
+      if (!hash || !hash[key]) {
+        return;
+      }
+      if (relationship.kind === 'hasMany' && isArray(hash[key])) {
+        forEach.call(hash[key], function(hash) {
+          if (hash) {
```

Comment:
`hash` as an argument name is a little confusing because there already is a `hash` argument from the parent scope. Lets rename this to something else. Maybe `error`?